LaMetric Time Binding

syphr42 opened the pull request:
New binding for the LaMetric Time Network-Connected Clock
http://lametric.com/

This binding aims to support all of the device features as well as native, third party, and owner-created apps with full data push support for owner-created apps.

Details can be found here at #1704

syphr42 commented:
This code will require several of the same changes suggested in #2009 before it is ready to be merged.

martinvw commented:
@syphr42 Please note that, because it's marked [WIP], I do and did not review this PR completely yet.

If it's finished and ready to review please update the title and mention me so I can remove the "Work in progress" label.

syphr42 commented:
@martinvw Thanks for the reminder. I've been distracted with other projects lately, but I hope to button this up soon so we can get it merged in. As it stands, it is still WIP.

syphr42 commented:
I rebased the PR to latest development and refactored some parts to bring it in alignment with current development and static analysis results. I also squashed all of the commits to prepare for review. There are still a few changes to be made, but this get's the PR back up to date.

kaikreuzer commented:
@syphr42 Just as a heads-up: I bought myself a LaMetric, it's really a neat device for openHAB purposes! I won't find any time to play much with it before the end of this year, but please expect me to come back to this PR some time after christmas! So if there is anything you'd want to clean up before I dive into it, now would be a good time to do so 😎

kaikreuzer commented:
One further observation: If I unplug my LaMetric and keep sending commands from openHAB to it, this results in those error logs (after a 1 minute timeout):
```
2017-12-27 18:10:50.375 [ERROR] [.i.c.AbstractInvocationHandler:101  ] - An error occurred while calling method 'ThingHandler.handleCommand()' on 'org.openhab.binding.lametrictime.handler.LaMetricTimeHandler@73ef43b': java.net.ConnectException: Operation timed out (Connection timed out)javax.ws.rs.ProcessingException: java.net.ConnectException: Operation timed out (Connection timed out)
	at org.glassfish.jersey.client.internal.HttpUrlConnector.apply(HttpUrlConnector.java:287)
	at org.glassfish.jersey.client.ClientRuntime.invoke(ClientRuntime.java:255)
	at org.glassfish.jersey.client.JerseyInvocation$1.call(JerseyInvocation.java:684)
	at org.glassfish.jersey.client.JerseyInvocation$1.call(JerseyInvocation.java:681)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:228)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:444)
	at org.glassfish.jersey.client.JerseyInvocation.invoke(JerseyInvocation.java:681)
	at org.glassfish.jersey.client.JerseyInvocation$Builder.method(JerseyInvocation.java:437)
	at org.glassfish.jersey.client.JerseyInvocation$Builder.put(JerseyInvocation.java:326)
Caused by: java.net.ConnectException: Operation timed out (Connection timed out)
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at sun.security.ssl.SSLSocketImpl.connect(SSLSocketImpl.java:668)
	at sun.net.NetworkClient.doConnect(NetworkClient.java:175)
	at sun.net.www.http.HttpClient.openServer(HttpClient.java:432)
	at sun.net.www.http.HttpClient.openServer(HttpClient.java:527)
	at sun.net.www.protocol.https.HttpsClient.<init>(HttpsClient.java:264)
```
Instead of this, I would simply expect the Thing to go OFFLINE.

syphr42 commented:
@kaikreuzer Sorry for the long delay - I've had a lot going on recently and haven't been able to get back to this. That said, I'm planning to devote some time this weekend so hopefully I will have all of your comments resolved soon.

kaikreuzer commented:
Sounds great, looking forward to moving it forward together!

syphr42 commented:
@kaikreuzer 

You made a lot of great suggestions and I've implemented several of them. However, the idea of removing generic apps and custom notifications is troubling to me. While the pre-installed core apps are a useful base (e.g. clock, weather, radio), the real power of this device is in its rich ecosystem of third-party apps, as a notification device, and the owner has the ability to write custom apps to which updates can be pushed. If we remove all that functionality it becomes not much more than a Squeezebox with weather and a timer. While I understand that passing JSON around isn't ideal, I would much rather keep what we have until there is a better replacement.

Given the above, I was looking into implementing your suggestion about ditching the "activate" buttons and having a single list of apps where the user could select one to activate. I love this idea and would much prefer it over the current button approach. However, I don't see a way to implement that without knowing the full list of apps at compile time (which is impossible with third-party apps). Do you have an implementation idea or were you thinking of this in the context of removing third-party app support only?

Also, regarding the package name config parameter, my goal was to keep configuration outside the code as much as possible. For generic apps, the package name must be a configuration property. For the provided LaMetric apps, we could hard code the package names inside the handler classes, but this feels like bad coding. Also, if LaMetric would ever push an update that changes those package names, at least users could override the values until a new version of the binding is created. I see no downside to having them as external properties with default values. Do you feel strongly about this change?

kaikreuzer commented:
Thanks for your updates @syphr42!

> the real power of this device is in its rich ecosystem of third-party apps, as a notification device

Yes, I fully get this and I am not trying to tell you that those should not be supported. What I am asking for is that we only merge code into the official repo, which must not be considered a hack (which the JSON channels clearly are). And no, you cannot convince me to first merge the hack in the hope that it's going to be soon replaced by a clean solution. I am old enough to know that this will never happen in reality ;-). So please let us do the very initial version of the binding without those channels and let us work together on proper rule actions, ideally by discussing the right architecture on https://github.com/openhab/openhab2-addons/pull/2926 first.

>  I don't see a way to implement that without knowing the full list of apps at compile time

There is no need to know them at compile time. All you need to do is to implement a [`DynamicStateDescriptionProvider`](https://github.com/eclipse/smarthome/blob/4204ce06bb28c28e5f711e720f87ef83beff2e27/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/type/DynamicStateDescriptionProvider.java#L30) and register it as an OSGi service. This provider can then provide a list of state options that are available.

> we could hard code the package names inside the handler classes, but this feels like bad coding.

No, that simply is "coding" and not "scripting" :-)

> if LaMetric would ever push an update that changes those package names

That is rather hypothetical. And we have that issue in general: If any device manufacturer decides to break its API, we have to adapt the binding and provide a new version.

> Do you feel strongly about this change?

Yes, I do, sorry 😇 . If there is no real need to expose internals, we shouldn't do it.

magcode commented:
Sorry for disturbing your tech discussions.
This is the voice of a "normal lametric user". The ONLY use case for me is currently to send alert/info notifications and indicator app data to my lamatric time. This works great using JSON. Just to let you know that without those features the binding will get useless for me (and I will continue with the old/snapshot binding version).

Sorry again for stepping in.

reyem commented:
>  I am old enough to know that this will never happen in reality ;-). So please let us do the very initial version of the binding without those channels and let us work together on proper rule actions, ideally by discussing the right architecture on #2926 first.

I agree with you on an overall OH architecture view that this is not a very nice interface. However, the comment of @magcode shows the user reality. I'm in the same boat and would have to use the snapshot build forever. If we go the road to wait for an OH Architecture change to provide support for indicator apps, this would be really bad. We also see on the example of #1255, on how long such a change takes to really become reality.  Limiting the functionality of the bindings because of platform architecture gaps is a bad practice as well as it stops innovation and end users on OH2. Therefore I beliefe the JSON interface is perfectly valid "workaround" with the capabilities the platform **currently** provides. If there would be a better way already now, I would be very happy to change it. But removing it, respectively not providing it, is a no go.

kaikreuzer commented:
@magcode & @reyem: The perfect workaround for the time being is to continue using an unofficial version, just as you do now.

Feature creep and and architecture decay are the no-go, as they prevent a project of this size to survive.

And honestly, if you check my https://github.com/openhab/openhab2-addons/pull/2926#issuecomment-357972671, you will see that there isn't really any major gap, it just needs to be done. So feel free to involve yourselves and drive it forward - the solution will be much nicer for any user.

reyem commented:
I see, we will not get that through ;-) too bad

So then let's at least design how such an action approach based on the rules engine should look such that we can provide an official solution asap:

Currently I have for example following code in my rule to update inside and outside temperature on my custom indicator app:

` val insideTemp = (LivingRoomTemp.state as DecimalType).floatValue
     val outsideTemp = (ASH2200Temp1.state as DecimalType).floatValue

     val jsonFrames = String::format('
        {"frames" : [
             {
                "icon":"i4286",
                "text":"%1$.1f °C"
             },
             {
                "icon":"i4285",
                "text":"%2$.1f °C"
             }
        ]}', insideTemp, outsideTemp )
 
 sendCommand(LametricKitchenMyTempApp, jsonFrames) `

whereas the important item is

`String LametricKitchenMyTempApp "Temperature App" {channel="lametrictime:genericApp:kitchen:myTemperatureApp:frames"}`

If I look now at the rules engine, still somehow the JSON format would have to be passed/configured, as this is the interface of the indicator app and cannot be known from the binding in advance. 
* Would that then be just kind of a JSON template configured with the action? I somehow did not yet understand why that should be cleaner than setting an item state. 
* So far I have just found simple String examples. But I haven't really worked yet with the new rule engine. Can we somehow provide structures for the user to use? For example provide classes for the different frame types, such that an UI just can ask for the icon and text in above example? and the number of frames would simply be achieved by adding frames to a list via a UI?

Which is the best information base about how the rule engine is working? Is there some more information than what is documented on?: https://www.eclipse.org/smarthome/documentation/features/rules.html

kaikreuzer commented:
Let's not look specifically at the new rule engine, but at how an "action" for scripts actually should look like.
Your example above is perfect to show how things should NOT be: JSON is not a format to be used within code/scripts, but it is a serialization format for objects that are used in such scripts.
Furthermore, we should never ever confront an average openHAB user with such JSON strings!

So ideally, as a user you would wish to be able to do this:
```
sendToLaMetric("myLMT1", "LametricKitchenMyTempApp", insideTemp, outsideTemp)
```
i.e. more abstract
```
sendToLaMetric(thingID, appID, params...)
```
The knowledge about the right jsonFrames for the appID should be internal to the binding - ideally, because you as a developer/tech-user of the custom app have already contributed it to the binding's source code (we could put those in a subfolder with the naming `<appID>.json`).

I understand that you also want to have the flexibility to use apps that are not already supported by the binding (as they are your own and not even published anywhere). Well, we can easily allow to feed those as additional configuration to the binding, which could simply be put in `conf/services/lametrictime.cfg`.

Would that suit your needs?

reyem commented:
The sample was a bit to simple. Unfortunately, LaMetric is designed differently. The apps you can publish are "Internet Apps". These are nicely configurable by the user, but poll stuff from the Internet or allow to push stuff to the cloud to be received by all users. So if I would publish my temperature frame app, anybody can see my temperatures on their lametrics, which does not really make sense. And this is the reason, why users will create their own private indicator apps, which are just reachable in the local network on the device. Thus, we cannot pre deliver the matching json definitions. An as @syphr42 might confirm, the json structure is also not published via API on the device.

(@syphr42 , @magcode : if I misunderstood this, please correct!)

following frames are theoretically possible:

Text Frame:
```
{
    "text": "Simple Text Frame $1, $2 whatever",
    "icon": null
}
```
goal frame:
```
{
    "goalData": {
    "start": 0,
    "current": 99,
    "end": 120,
    "unit": "kg"
    },
    "icon": "i120"
}
```

chart frame:

```
{
    "index": 3,
    "chartData": [
           1,
           3,
           4,
           5,
           7,
           4,
           2
       ]
}
```

Now, to support them fully, we would need to pass kind of structures to an api.

Taking your proposal we would need something to create structure to pass as arguments, at least for goals or charts. So maybe it could look like

```
var textFrame = LaMetricUtils.createTextFrame(null, String::format( "Simple Text Frame $1, $2 whatever", item1, item2 ) );
var chartFrame = LaMetricUtils.createChart("i120", 1,2,3,4,5,6,7,4,2);
var goalFrame = LaMetricUtils.createGoal(0,29,120, "kg");
sendToLaMetric("myLMT1", "LametricKitchenMyTempApp", textFrame, chartFrame, goalFrame);
```

Now I just ask:

1. is this really easier?
2. the user is anyways confronted with the json structure in this case as he "designs" the app via https://developer.lametric.com/applications and copies the format from there. imho abstracting it in OH leads rather to confusion. So why would we do that?

@kaikreuzer : As you also own a device, have you tried using it to simply shows something from your OH items on the device?

However, limiting the API to text frames without icons, we could actually provide a simple action like the already proposed:

```
sendToLaMetric(thingID, appID, textFrameContents...)
```
where each textFrameContent parameter results in its own text frame.
Important to know about indicator apps is, that one can actually send different frames to the app than the app defined. So if I define an app with 2 chart frames, it also accepts 2 text frames. 

Still, it would be bad to limit to just text frames. goals can be a nice visualization for things like thermostats ( current temperature vs target temperature). And also with charts, trends could be visualized. 

An other idea I had in the really start of the binding, was to actually use something like the OH sitemaps in a limited form. Sitemaps would interestingly already contain all required structural elements with SetPoint -> Goal, Chart -> Chart, Text -> Text. Also frames are there. But of course they would need to be limited. But on a high abstraction level, LaMetric is actually a User Interface... Maybe this is another path to think, although that might probably change the binding quite a bit...

magcode commented:
Some thoughts...
>https://www.eclipse.org/smarthome/documentation/features/rules.html

I was also reading it. Rather abstract and hard to understand. I gave up.

> is this really easier?

Yes. A bit. I reduces the risk of broken JSON structures.

>indicator apps

Don't focus too much on the indicator apps. If you look at notifications (http://lametric-documentation.readthedocs.io/en/latest/reference-docs/device-notifications.html) ... there is _nothing_ predefined in any way. So one is completely free what the device should notify. One of my rules as an example:

```
rule "LametricWindowAlertrule"
    when
        Item LametricWindowAlert received update
    then
    val jsonMessage = String::format('{
                "priority": "critical",
                "icon_type" : "alert",
                "model": {
                    "cycles": 1,
                    "frames": [
                    {
                        "icon": "a135",
                        "text":"%1$s"
                    }],
                    "sound": {
                        "category": "notifications",
                        "id": "negative4",
                        "repeat": 3
                    }
                }
            }', LametricWindowAlert.state.toString)             
    LametricAdvancedNotification.sendCommand(jsonMessage)
end
```

It will be hard to put that into some user friendly / easy to use configuration. So I'm currently in favor of the `LaMetricUtils.createTextFrame` API/helper style.

Btw. look how others are doing it: https://home-assistant.io/components/notify.lametric/

kaikreuzer commented:
> the user is anyways confronted with the json structure in this case as he "designs" the app

You need to differentiate between the user groups. YOU are the administrator/developer. Your spouse, kid, father, friend or customer for whom you have set up the system is the user. I can succeed to tell my wife to create a rule using the Paper UI that sends a new kind of notification to my custom app. I won't even try to tell her about how to write and debug JSON strings correctly.

And as I said: Not even JS developers will feel tempted to write JSON in the script. After all, they directly write JS objects in code - and this is what they would e.g. also do when using JSR223 rules.

> It will be hard to put that into some user friendly / easy to use configuration.

I agree, but does it really make sense to put that in the rule? If I read your example correctly, it is a completely static JSON string and you are not at all constructing it dynamically in that rule. What is worse, you cannot even re-use it in a similar rule that might want to use the same notification style, but just pass another value. Wouldn't it be much more elegant to put that JSON string as a property in `conf/services/lametrictime.cfg`(as suggested above) and only reference it from the rule and pass the required parameter to the call?

> Btw. look how others are doing it

This seems to prove my point: Start simple, don't bother the user (not the admin!) with JSON.

syphr42 commented:
I haven't had a chance to dive into the rule discussion yet - I'll get to that soon. However, in the meantime I implemented a new String channel type on the device called `app` that takes an application widget reference in the form `package name:widget ID` to activate that application. I also implemented the `DynamicStateDescriptionProvider` to generate valid states automatically from the installed applications on the device.

One thing I noticed when working on this change was that the Basic UI driven off a sitemap does not seem to recognized the configured states (selections or switches). Therefore the selection dropdown is empty. However, the Paper UI picked up on the dynamic states. If I get a chance, I will look into why that is happening as well. Due to this deficiency, I don't believe the `activate` channel on the apps should be removed yet. For now, I just marked them as advanced.

kaikreuzer commented:
@syphr42 Great that you managed to do the dynamic state descriptions!

Wrt the Basic UI, you are right - users currently have to define a mapping in their sitemap manually for that (but it has always been that way for any kind of selections). But just this week, exactly your request came up as well, see https://github.com/eclipse/smarthome/issues/4942.

> I don't believe the activate channel on the apps should be removed yet. For now, I just marked them as advanced.

No, please do not keep them for the same reasons as above: Although users have to define the mapping in the sitemap manually, that's simply how openHAB works so far, please don't fight it, but rather help on https://github.com/eclipse/smarthome/issues/4942. Once channels are in the binding, you cannot easily remove them again as it would be a breaking change - so don't let us introduce stuff that we know needs to be removed (and anyhow: Using these channels in a sitemap is actually much more work than defining the mapping, isn't it?).

Wrt the discussion above, would you also remove the genericApp thing type from this PR?

reyem commented:
regarding the configuration / action. I think you guys convinced me with the idea of the configuration of the json "template". This could be used for both, notifications and indicator apps. 
However, in the case of indicator apps it would even be better if the binding could take care of updating the apps in case of the item change. Currently I need to create a rule to get the updated temperature on the indicator apps. 
Therefore the question, can a binding actually "listen" to item update events?

syphr42 commented:
Since we are planning to make some deep cuts in functionality (at least temporarily) to this binding until we build out a more robust solution for generic apps and notifications, I have created a new branch on my fork that contains the current state of the binding. Please use this if you need the current functionality until we bring it back into the official binding.

https://github.com/syphr42/openhab2-addons/tree/lametric-binding-with-generic-apps

kaikreuzer commented on addons/binding/org.openhab.binding.lametrictime/src/main/java/org/openhab/binding/lametrictime/internal/LaMetricTimeStateDescriptionProvider.java:
```diff
@@ -0,0 +1,42 @@
+@Component(service = { DynamicStateDescriptionProvider.class,
+        LaMetricTimeStateDescriptionProvider.class }, immediate = true)
+public class LaMetricTimeStateDescriptionProvider implements DynamicStateDescriptionProvider {
+    private final Map<ChannelUID, StateDescription> stateDescriptions = new ConcurrentHashMap<>();
```
You should do a `clear()` of that map within `deactivate()` of this component.

syphr42 replied:
Will do.

kaikreuzer commented on addons/binding/org.openhab.binding.lametrictime/src/main/java/org/openhab/binding/lametrictime/internal/LaMetricTimeStateDescriptionProvider.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Override
+    public @Nullable StateDescription getStateDescription(@NonNull Channel channel,
+            @Nullable StateDescription originalStateDescription, @Nullable Locale locale) {
```
You receive the originalStateDescription as a parameter here - imho, you should not plainly overwrite it, but rather use its content and only fill in the options, which is what you are for.

syphr42 replied:
I'll re-work this so only the state options are used. I was trying to allow more functionality, but it shouldn't be hard to limit it to just options.

kaikreuzer commented on addons/binding/org.openhab.binding.lametrictime/src/main/java/org/openhab/binding/lametrictime/internal/LaMetricTimeUtil.java:
```diff
@@ -0,0 +1,51 @@
+            return coreAppLabel;
+        }
+
+        return "Generic App by " + app.getVendor();
```
This looks actually quite nasty on the app option list in the UI. To improve the readability of the list, maybe `app.getVendor() + "'s App"` would be a better option?

syphr42 replied:
Sure, I'll change the format to the one you suggested.

kaikreuzer commented on addons/binding/org.openhab.binding.lametrictime/src/main/java/org/openhab/binding/lametrictime/internal/discovery/LaMetricTimeAppDiscoveryService.java:
```diff
@@ -0,0 +1,131 @@
+            }
+        }
+
+        stopScan();
```
you shouldn't call stopScan yourself

syphr42 replied:
I'll remove this.

kaikreuzer commented on addons/binding/org.openhab.binding.lametrictime/src/main/java/org/openhab/binding/lametrictime/internal/discovery/LaMetricTimeAppDiscoveryService.java:
```diff
@@ -0,0 +1,131 @@
+    @Override
+    protected void startBackgroundDiscovery() {
+        logger.debug("Starting background discovery for new apps");
+        startScan();
```
A background discovery is supposed to run regularly and not just once. I am not sure though, whether we really need a regular scan as usually not much will change (and it isn't hard to retrigger a scan for the user, in case he has added a new app).
So I'd rather suggest to remove this method and pass a `false`in line 71.

One issue I noticed anyhow: The auto-discovery of the apps didn't work, I had to manually run a scan to make them appear. I assume this is because the call here happens before the handler has connected to the device.

It therefore might make sense to introduce an AppListener interface, with which the discovery service can register as a listener at the handler, so that the handler can pass all found apps to the listener once it is connected. Wdyt?

syphr42 replied:
It would probably be better to keep this simple and just remove the background scan letting the user hit the button to pick up new apps. The manual scan is working fine and like you said, it doesn't seem that hard for the user to hit the scan button when new apps are added that he/she would like to access in OH. An automatic scan for new apps can be an improvement later if there is demand, especially since we are removing generic app support entirely for now.

syphr42 commented on addons/binding/org.openhab.binding.lametrictime/ESH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,85 @@
+        <item-type>String</item-type>
+        <label>Application</label>
+        <description>Active application</description>
+        <state>
```
I tested out removing this, but it causes the Paper UI to fail to render the item as a selection. With the placeholder in there, everything works as expected. Perhaps there is a timing issue here. This might require some additional investigation if we really want to remove it, but it doesn't seem like a big deal to leave it.

kaikreuzer replied:
I think there must be some timing issue. I see this effect from time to time even if this xml is in place (in that case it shows the invalid option "App", which isn't imho any better, so I'd still vote for removing this part here).

syphr42 replied:
Ok; I'll remove it.

syphr42 commented on addons/binding/org.openhab.binding.lametrictime/ESH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,85 @@
+        <description>Perform an action with a simple trigger (no parameters)</description>
+    </channel-type>
+    
+    <channel-type id="controlType">
```
Will do.

syphr42 commented on addons/binding/org.openhab.binding.lametrictime/ESH-INF/thing/device.xml:
```diff
@@ -0,0 +1,33 @@
+            <channel id="volume" typeId="volumeType"/>
+            <channel id="bluetooth" typeId="bluetoothActiveType"/>
+            <channel id="app" typeId="appType"/>
+            <channel id="info" typeId="simpleNotificationType"/>
```
Will do.

syphr42 commented on addons/binding/org.openhab.binding.lametrictime/ESH-INF/thing/device.xml:
```diff
@@ -0,0 +1,33 @@
+            </parameter>
+            <parameter name="apiKey" type="text" required="true">
+                <label>API key</label>
+                <description>API key to access LaMetric Time</description>
```
Will do.

syphr42 commented on addons/binding/org.openhab.binding.lametrictime/ESH-INF/thing/weather-app.xml:
```diff
@@ -0,0 +1,28 @@
+        <channels>
+            <channel id="forecast" typeId="triggerActionType">
+                <label>Forecast</label>
+                <description>Show the forecast</description>
```
The device shows the current conditions all the time when you have the weather app activated. It shows the forecast momentarily when you send a request for it. Therefore, the on/off pattern doesn't really work. You could add a rule to automatically turn off the forecast switch as soon as you press it so that it acts more like a momentary switch than a toggle.

The main problem with a device like LaMetric Time and OH is that OH expects channels to have state and the LaMetric Time doesn't provide a lot of status information. For example, it doesn't say which app is active or what the clock alarm is set to or which station is playing on the radio. Most of the interaction is one-way towards the device. It would be nice if we had something like trigger channels, but in reverse. Something that simply sends information to a thing and expects nothing in return and no state information to be provided or retained in OH.

kaikreuzer replied:
Yes, so here I would suggest a similar solution as mentioned above: Instead of a Switch, use a String channel and define "FORECAST" as the only option for it. Once we have https://github.com/eclipse/smarthome/issues/4942 in place, this will nicely render a single button for this (and not showing a toggle switch). One day we might also have a solution for https://github.com/eclipse/smarthome/issues/595, so that there won't be any state changes for that channel, so that it effectively becomes a "stateless command only channel".

syphr42 replied:
I'll change this and the "stopAlarm" channel as well.

syphr42 replied:
I found an issue with the command channels. If there is only one option (e.g. weather forecast), once you set it, you can't use it again because there is no way to change the state. To avoid this, I'm clearing the state in the handler using `updateState(channelUID, new StringType())`. This works well for the Basic UI, but the Paper UI doesn't get an update to clear out the state value. If you refresh the Paper UI, it does have the updated empty value.

Do you consider this a bug in the Paper UI or is it wrong for me to change the state while handling the command?

kaikreuzer replied:
Both :-)
The Paper UI should react on such state changes, so I'd consider it a bug (@htreu, would you agree?) in the Paper UI.
Doing such a state change shouldn't be necessary, but the issue is that there is currently no way to suppress the auto-update (see https://github.com/eclipse/smarthome/issues/595), thus you cannot prevent the stateless "command channel" to suddenly have a state.
This will hopefully be addressed soon through https://github.com/eclipse/smarthome/pull/5011.

syphr42 commented on addons/binding/org.openhab.binding.lametrictime/src/main/java/org/openhab/binding/lametrictime/LaMetricTimeBindingConstants.java:
```diff
@@ -0,0 +1,64 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
```
Will do.

syphr42 commented on addons/binding/org.openhab.binding.lametrictime/ESH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,85 @@
+                          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                          xmlns:thing="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0"
+                          xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
+    <channel-type id="jsonNotificationType">
```
This will be removed until we figure out the replacement action.

syphr42 commented on addons/binding/org.openhab.binding.lametrictime/ESH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,85 @@
+        </state>
+    </channel-type>
+    
+    <channel-type id="framesType" advanced="true">
```
This will be removed until we figure out the replacement action.

syphr42 commented on addons/binding/org.openhab.binding.lametrictime/ESH-INF/thing/device.xml:
```diff
@@ -0,0 +1,33 @@
+            <channel id="info" typeId="simpleNotificationType"/>
+            <channel id="warning" typeId="simpleNotificationType"/>
+            <channel id="alert" typeId="simpleNotificationType"/>
+            <channel id="advanced" typeId="jsonNotificationType"/>
```
This will be removed until we figure out the replacement action.

syphr42 commented on addons/binding/org.openhab.binding.lametrictime/ESH-INF/thing/generic-app.xml:
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
I was thinking more about the generic apps - there is some value in being able to at least select them as the active app on the device. What do you think about leaving them in, but removing the `frames` channel for now?

syphr42 replied:
Never mind what I said here. When implementing the state provider, I was considering using the things rather than the apps actually on the device, but I scrapped that idea. We can remove this without affecting the selection list.

syphr42 commented on addons/binding/org.openhab.binding.lametrictime/about.html:
```diff
@@ -0,0 +1,48 @@
+    <br/>
+    <pre>Apache 2.0 License</pre>
+    </br>
+    Source: https://github.com/syphr42/liblametrictime-java
```
On this topic specifically, I feel quite strongly that the device API should remain separate. While there may not be any public users on GitHub, I am using this API in other projects. But more importantly, I very much believe in software reusability and would argue against this sort of bundling. Ideas like the Unix philosophy of doing one thing and doing it well and the importance of low coupling and high cohesion in software units are important to me. I also believe strongly in the Apache license and would like as much of my code as possible to be under that license. 

Since the API source is Apache 2.0 licensed, you are free to copy the code into OH if you choose to, but my vote is against it.

kaikreuzer replied:
Ok, understood and fine for me.
My only wish being that you do your best to keep the code base clean wrt intellectual property, i.e. that you make sure to have potential future contributors sign [the DCO or an CLA](https://julien.ponge.org/blog/developer-certificate-of-origin-versus-contributor-license-agreements/) as it is best practice in open source projects.

syphr42 replied:
That's fair. I added the [DCO](https://github.com/syphr42/liblametrictime-java/blob/master/CONTRIBUTING.md) and installed a [DCO check](https://github.com/probot/dco).

syphr42 commented on addons/binding/org.openhab.binding.lametrictime/ESH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,85 @@
+        <description>Perform an action with parameters as input</description>
+    </channel-type>
+    
+    <channel-type id="triggerActionType">
```
Will do.

syphr42 commented on addons/binding/org.openhab.binding.lametrictime/ESH-INF/thing/countdown-app.xml:
```diff
@@ -0,0 +1,40 @@
+                <label>Configure</label>
+                <description>Configure the timer</description>
+            </channel>
+            <channel id="pause" typeId="triggerActionType">
```
Unfortunately, we can't really combine these. The device does not provide state as to whether it is paused or not.  Sending `start` while it is running does nothing. Sending `pause` while it is paused, does nothing. Without separating these, we can easily become out of phase (due to other input, like the mobile app or the device itself). If that happens, the user has to flip the switch more than once to see any effect. Keeping them separate means the user can choose the exact functionality required.

kaikreuzer replied:
Hm, it is really unfortunate that one cannot get the current states from the device.
Nonetheless, I would claim that combining those two channels makes sense, since at least, people can use a toggle switch, if they only control it remotely (and by toggling twice they will in any case get the intended result).

Anyhow, it does not make sense to have two commands (ON and OFF) for every single trigger here. If we say that the state does not make any sense (as it gets out of sync too easily), we should actually not model any separate channels at all, but rather have a single "command channel", which would take Strings like PLAY/PAUSE/RESET (and the user could have a single button for each of those options in the UI).

Which way would you prefer?

syphr42 replied:
I like the idea of a command channel - the toggles always felt clumsy to me. I'd love to see something like a momentary button in the future (like a trigger channel in reverse) as part of the core, but this is probably a good solution for now. 

I'll try this out. Let me know what you think of the result.

syphr42 commented on addons/binding/org.openhab.binding.lametrictime/ESH-INF/thing/stopwatch-app.xml:
```diff
@@ -0,0 +1,36 @@
+        <description>Provides control for the stopwatch app on the LaMetric Time</description>
+        
+        <channels>
+            <channel id="pause" typeId="triggerActionType">
```
See comment above on the same topic for the countdown app.

kaikreuzer replied:
See also my comment from above.

syphr42 commented on addons/binding/org.openhab.binding.lametrictime/ESH-INF/thing/countdown-app.xml:
```diff
@@ -0,0 +1,40 @@
+        <description>Provides control for the countdown app on the LaMetric Time</description>
+        
+        <channels>
+            <channel id="configure" typeId="parameterActionType">
```
To provide some value for now, rather than delete this completely I will change it to a duration (in seconds) so it can be a simple Number channel. The only thing the user can't do is set and start the countdown in one action now, but all of the functionality is still there.

syphr42 commented on addons/binding/org.openhab.binding.lametrictime/ESH-INF/thing/clock-app.xml:
```diff
@@ -0,0 +1,32 @@
+        <description>Provides control for the clock app on the LaMetric Time</description>
+        
+        <channels>
+            <channel id="setAlarm" typeId="parameterActionType">
```
This is a little awkward since the date is ignored and only time is sent. I'll give this a shot. Let me know what you think when you see it.

kaikreuzer replied:
I agree that this is awkward, but it isn't the fault of the binding, but rather the framework - I'd suggest to leave it here as is and created https://github.com/eclipse/smarthome/issues/5001 to improve the situation.

syphr42 replied:
:+1:

syphr42 commented on addons/binding/org.openhab.binding.lametrictime/ESH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,85 @@
+        <description>Update app frames using JSON</description>
+    </channel-type>
+    
+    <channel-type id="parameterActionType" advanced="true">
```
Will do.

kaikreuzer commented:
Just ftr, as I saw it in my log:
```
2018-02-01 14:45:34.284 [INFO ] [s.event.ItemStateChangedEvent :51   ] - lametrictime_device_SA170401346100W00BS9_app changed from NULL to reference
2018-02-01 14:45:34.304 [DEBUG] [.l.handler.LaMetricTimeHandler:151  ] - Unexpected error while handling command - taking clock offlinejava.lang.IllegalArgumentException: Provided string 'reference' is not in the format '<package name>:<widget ID>'
	at org.openhab.binding.lametrictime.internal.WidgetRef.fromString(WidgetRef.java:29)
	at org.openhab.binding.lametrictime.handler.LaMetricTimeHandler.handleAppCommand(LaMetricTimeHandler.java:203)
	at org.openhab.binding.lametrictime.handler.LaMetricTimeHandler.handleCommand(LaMetricTimeHandler.java:141)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:153)
	at org.eclipse.smarthome.core.internal.common.InvocationHandlerSync.invoke(InvocationHandlerSync.java:59)
	at com.sun.proxy.$Proxy103.handleCommand(Unknown Source)
	at org.eclipse.smarthome.core.thing.internal.profiles.ProfileCallbackImpl.handleCommand(ProfileCallbackImpl.java:72)
```

@syphr42 Did you see my other inline comments from 3 days ago? Just asking because they are hidden/collapsed in the GIthub view, so you might have missed them.

syphr42 commented:
Yes, I did see your comments, but I haven't had a chance to address them
yet. I should have time this weekend. I'll look at the error you got as
well on the new app channel.

syphr42 commented:
@kaikreuzer I believe I have addressed all of your concerns. Also, the stacktrace you saw in the log was the placeholder state option for the app channel. I've removed that placeholder now so you shouldn't see that issue again.

kaikreuzer commented:
Many thanks @syphr42!
Code looks good and I have just done some further testing - I directly pushed by findings as new commits to your branch, please check that these are ok.

One last issue that would be nice to address before merging: It seems that the HTTP requests have a terribly long timeout (60s?). This makes the OFFLINE detection awkwardly slow as you only learn a minute after sending a request that it wasn't successful. Imho, setting the timeout to 10 seconds should be long enough. Unfortunately, all this code is encapsulated in your external library, so I didn't have an easy way to fix this here. Could you provide an updated version?

syphr42 commented:
@kaikreuzer 

Your changes look good. I do have one question though: why did you make the Bluetooth channel advanced? Essentially you flip the switch to turn it on and then pair your phone or some other device to start streaming. I just want to know for future reference to help me make the decision about whether a channel should be advanced or not.

Also, regarding your comment about the timeout, I've added the ability to pass a ClientBuilder into the API when creating an instance so that custom properties can be set (in this case `ClientProperties.CONNECT_TIMEOUT` and `ClientProperties.READ_TIMEOUT`). I just performed a release on the API to Maven Central so I will take that official version 0.1.0 and use it to update the binding. I should have an update pushed tomorrow night.

kaikreuzer commented:
> why did you make the Bluetooth channel advanced? 

Well, this can always be negotiated.
As a rule of thumb, the number of non-advanced channels should be kept small and only to the channels that most likely most people will want to use right away.
If I understand this feature correctly, you can keep Bluetooth enabled and are at any time free to use the device as a bluetooth speaker. So the feature toogle the Bluetooth support did not seem to be a regular/daily use case to me. If I am wrong here, feel free to remove the advanced flag again.

syphr42 commented:
I hadn't really thought about users leaving the Bluetooth enabled - I've always turned it off when I'm not using it. I'm assuming I'm in the minority there so your explanation makes sense. Let's leave it as advanced.

kaikreuzer commented:
Thanks for the update, I'd say we are good to go for the initial version!
